repository-data/webfiles/src/main/resources/site/src/js/sticky-nav/sticky-nav.js

```diff
@@ -1,41 +1,143 @@
-    const CHAPTER_NAV_SPACING = 24;
-    const SIDEBAR_SPACING = 20;
+function addMarker(elm) {
+    let elementById = document.getElementById(`#${elm.id}_list`)
+    if (elementById !== null) {
+        let markers = elementById.getAttribute(NAV_MARKER);
+        if (markers !== null) {
+            elementById.setAttribute(NAV_MARKER, markers + "|");
+        } else {
+            elementById.setAttribute(NAV_MARKER, "|");
+        }
+    }
+}
+
+function removeMarker(elm) {
+    let elementById = document.getElementById(`#${elm.id}_list`)
+    if (elementById !== null) {
+        let markers = elementById.getAttribute(NAV_MARKER);
+        if (markers !== null && markers.length > 1) {
+            elementById.setAttribute(NAV_MARKER, markers.substring(1));
+        } else {
+            elementById.removeAttribute(NAV_MARKER);
+        }
 
-    return chapterNav ? (chapterNav.offsetHeight + (CHAPTER_NAV_SPACING + SIDEBAR_SPACING)) : SIDEBAR_SPACING;
+    }
 }
 
+/**
+ * Finds any elements marked with the CSS class 'navigationMarker' and calculates
+ * whether it in or out of view - then marks/unmasks the corresponding sticky nav
+ * list items appropriately.
+ */
 export function initStickyNav() {
-    if (stickyNav) {
-        initScrollSpy();
-
-        const sidebar = new StickySidebar('#sticky-nav', {
-            topSpacing: getStickyOffset(),
-            bottomSpacing: 40,
-            resizeSensor: true,
-            containerSelector: '.grid-row'
-        });
-
-        $(window).on('load resize', function () {
-            if (window.matchMedia('(max-width: 924px)').matches) {
-                sidebar.destroy();
+    const io = new IntersectionObserver(entries => {
+        for (const entry of entries) {
+            if (entry.isIntersecting) {
+                addMarker(entry.target);
             } else {
-                sidebar.initialize();
+                removeMarker(entry.target);
             }
-        });
+        }
+    }, {threshold: 0})
 
-        $('.article-section-nav__list a').click(function () {
-            const match = jQuery(this).attr('href').match(/#\S+/);
-            const url = match ? window.location.pathname + match[0] : window.location.pathname;
-            logGoogleAnalyticsEvent('Link click', 'Section Nav', url);
-        });
-    }
+    document.querySelectorAll(`.${INDICATOR}`)
+        .forEach(el => io.observe(el))
+}
+
+
+/**
+ * Finds logically nested elements that are marked with the CSS class
+ * 'navigationMarker-sub' and calculates whether it's relevant logical parent is in
+ * or out of view - then marks/unmasks the corresponding sticky nav list items
+ * appropriately.
+ *
+ * See above for details on 'logically nested' elements.
+ */
+export function initStickyNavSubItems() {
+    const io = new IntersectionObserver(entries => {
+        const getPreviousRelevantSibling = (elem, selector) => {
+            let sibling = elem.previousElementSibling
+            while (sibling) {
```

Comment:
What's the function of `while` here? `previousElementSibling` should contain a single node or null.